File: Server~/src/resources/getProfilerResource.ts
import { Logger } from "../utils/logger.js";
import { McpUnity } from "../unity/mcpUnity.js";
import { ResourceTemplate, McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { McpUnityError, ErrorType } from "../utils/errors.js";
import { ReadResourceResult } from '@modelcontextprotocol/sdk/types.js';
import { Variables } from '@modelcontextprotocol/sdk/shared/uriTemplate.js';

// Constants for the resource
const resourceName = "get_profiler_data";
const resourceUriTemplate = "unity://profiler/{dataType}";
const resourceMimeType = "application/json";

/**
 * Registers the Unity Profiler data resource with the MCP server
 * This resource provides access to Unity's performance profiling data
 *
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerGetProfilerResource(
  server: McpServer,
  mcpUnity: McpUnity,
  logger: Logger
) {
  // Create a resource template with the MCP SDK
  const resourceTemplate = new ResourceTemplate(
    resourceUriTemplate,
    {
      list: undefined // No listing functionality needed for profiler data
    }
  );
  logger.info(`Registering resource: ${resourceName}`);

  // Register this resource with the MCP server
  server.resource(
    resourceName,
    resourceTemplate,
    {
      description: "📊 获取Unity实时性能数据，包含CPU使用率、内存统计、GPU帧时间、调用层次分析和性能建议。dataType参数用于过滤特定指标：'cpu'-CPU性能，'memory'-内存使用，'gpu'-GPU指标，'rendering'-渲染统计(Draw Calls/Batches)，'hierarchy'-调用堆栈分析，'summary'-性能概览，'all'-完整数据。当需要了解性能状况时优先使用此资源。",
      mimeType: resourceMimeType
    },
    async (uri: URL, variables: Variables): Promise<ReadResourceResult> => {
      try {
        logger.info(`Fetching profiler resource`, { uri: uri.toString() });
        
        // Extract dataType from URI template variables
        const dataType = (variables["dataType"] as string) || "all";
        
        // Validate dataType parameter
        const validTypes = ["all", "cpu", "memory", "gpu", "rendering", "hierarchy", "summary"];
        const validatedDataType = validTypes.includes(dataType.toLowerCase()) ? dataType.toLowerCase() : "all";

        // Send request to Unity using the resource method name
        const response = await mcpUnity.sendRequest({
          method: resourceName,
          params: {
            dataType: validatedDataType
          }
        });

        if (!response.success) {
          throw new McpUnityError(
            ErrorType.RESOURCE_FETCH,
            response.message || "Failed to fetch profiler data from Unity"
          );
        }

        // Format the response data for better readability
        const formattedData = formatProfilerData(response.data || response, validatedDataType);
        
        return {
          contents: [
            {
              uri: `unity://profiler/${validatedDataType}`,
              mimeType: resourceMimeType,
              text: JSON.stringify(formattedData, null, 2)
            }
          ]
        };
        
      } catch (error) {
        logger.error(`Failed to fetch profiler resource`, error);
        throw error;
      }
    }
  );
}

/**
 * Format profiler data for better presentation to AI assistants
 * @param data Raw profiler data from Unity
 * @param dataType Type of data requested
 * @returns Formatted data with explanations and context
 */
function formatProfilerData(data: any, dataType: string): any {
  const formatted: any = {
    dataType: dataType,
    timestamp: data.timestamp || new Date().toISOString(),
    frameCount: data.frameCount || 0,
    data: data
  };

  // Add explanatory context based on data type
  switch (dataType) {
    case "cpu":
      formatted.explanation = {
        description: "CPU performance metrics showing time spent in different engine systems",
        metrics: {
          mainThreadFrameTime: "Time spent on main thread per frame (target: <16.67ms for 60fps)",
          behaviourUpdate: "Time spent in Update() methods across all scripts",
          physicsFixedUpdate: "Time spent in physics simulation",
          fps: "Current frames per second"
        },
        recommendations: generateCPURecommendations(data)
      };
      break;
      
    case "memory":
      formatted.explanation = {
        description: "Memory usage statistics for system and garbage collection",
        metrics: {
          systemUsedMemory: "Total memory used by the application (MB)",
          gcReservedMemory: "Memory reserved by garbage collector (MB)",
          gcUsedMemory: "Memory currently used by managed objects (MB)",
          totalAllocatedMemory: "Total allocated memory including native (MB)"
        },
        recommendations: generateMemoryRecommendations(data)
      };
      break;
      
    case "gpu":
      formatted.explanation = {
        description: "GPU performance metrics for rendering",
        metrics: {
          gpuFrameTime: "Time spent on GPU per frame (target: <16.67ms for 60fps)"
        },
        recommendations: generateGPURecommendations(data)
      };
      break;
      
    case "rendering":
      formatted.explanation = {
        description: "Rendering statistics including draw calls and geometry",
        metrics: {
          drawCalls: "Number of draw calls per frame (lower is better)",
          batches: "Number of batches after batching optimizations",
          setPassCalls: "Number of material/shader switches",
          triangles: "Total triangles rendered",
          vertices: "Total vertices processed"
        },
        recommendations: generateRenderingRecommendations(data)
      };
      break;
      
    case "hierarchy":
      formatted.explanation = {
        description: "CPU call hierarchy showing most expensive methods",
        metrics: {
          topCostlyMethods: "Methods taking the most CPU time, sorted by total time",
          frameIndex: "Frame number when data was captured",
          frameTimeMs: "Total frame time in milliseconds"
        }
      };
      break;
      
    case "summary":
      formatted.explanation = {
        description: "Performance summary with key metrics and automatic issue detection",
        metrics: {
          currentFPS: "Current frames per second",
          performanceGrade: "Overall performance grade (A-F)",
          issues: "Automatically detected performance issues",
          recommendations: "Suggested optimizations based on current metrics"
        }
      };
      break;
      
    case "all":
      formatted.explanation = {
        description: "Comprehensive profiler data including all metrics and analysis",
        sections: {
          cpu: "CPU performance metrics and timing",
          memory: "Memory usage and garbage collection stats",
          gpu: "GPU rendering performance",
          rendering: "Draw calls, batching, and geometry statistics",
          summary: "Overall performance assessment with recommendations"
        }
      };
      break;
  }

  return formatted;
}

/**
 * Generate CPU-specific recommendations based on the data
 */
function generateCPURecommendations(data: any): string[] {
  const recommendations: string[] = [];
  
  if (data.mainThreadFrameTime > 16.67) {
    recommendations.push("CPU frame time exceeds 60fps target. Consider optimizing Update() methods.");
  }
  
  if (data.behaviourUpdate > 8) {
    recommendations.push("High Update() overhead detected. Cache expensive operations and reduce update frequency.");
  }
  
  if (data.physicsFixedUpdate > 5) {
    recommendations.push("Physics taking significant CPU time. Consider reducing Rigidbody count or simplifying colliders.");
  }
  
  if (data.fps < 30) {
    recommendations.push("Very low FPS detected. Consider reducing scene complexity or implementing LOD system.");
  }
  
  return recommendations;
}

/**
 * Generate memory-specific recommendations based on the data
 */
function generateMemoryRecommendations(data: any): string[] {
  const recommendations: string[] = [];
  
  if (data.gcReservedMemory > 200) {
    recommendations.push("High GC memory usage. Implement object pooling to reduce allocations.");
  }
  
  if (data.systemUsedMemory > 1000) {
    recommendations.push("High system memory usage. Consider texture compression and audio optimization.");
  }
  
  const gcEfficiency = data.gcUsedMemory / Math.max(data.gcReservedMemory, 1);
  if (gcEfficiency < 0.7) {
    recommendations.push("Low GC efficiency. Avoid frequent allocations and use StringBuilder for string operations.");
  }
  
  return recommendations;
}

/**
 * Generate GPU-specific recommendations based on the data
 */
function generateGPURecommendations(data: any): string[] {
  const recommendations: string[] = [];
  
  if (data.gpuFrameTime > 16.67) {
    recommendations.push("GPU frame time exceeds 60fps target. Optimize shaders and reduce rendering complexity.");
  }
  
  return recommendations;
}

/**
 * Generate rendering-specific recommendations based on the data
 */
function generateRenderingRecommendations(data: any): string[] {
  const recommendations: string[] = [];
  
  if (data.drawCalls > 1000) {
    recommendations.push("High draw call count. Use static batching, GPU instancing, or texture atlases.");
  }
  
  if (data.setPassCalls > 500) {
    recommendations.push("High SetPass calls. Reduce number of unique materials.");
  }
  
  if (data.triangles > 1000000) {
    recommendations.push("High triangle count. Implement LOD system or reduce model complexity.");
  }
  
  if (data.drawCalls > 0 && data.batches > 0) {
    const batchingEfficiency = data.batches / data.drawCalls;
    if (batchingEfficiency < 0.5) {
      recommendations.push("Poor batching efficiency. Enable static batching in Player Settings.");
    }
  }
  
  return recommendations;
} 

File: Editor/Resources/GetProfilerResource.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using UnityEditor;
using UnityEditor.Profiling;
using UnityEditorInternal;
using Unity.Profiling;
using Newtonsoft.Json.Linq;

namespace McpUnity.Resources
{
    /// <summary>
    /// Resource for retrieving Unity Profiler performance data
    /// </summary>
    public class GetProfilerResource : McpResourceBase
    {
        private readonly Dictionary<string, ProfilerRecorder> _profilerRecorders = new Dictionary<string, ProfilerRecorder>();
        private readonly List<int> _parentsCacheList = new List<int>();
        private readonly List<int> _childrenCacheList = new List<int>();
        
        public GetProfilerResource()
        {
            Name = "get_profiler_data";
            Description = "Retrieves Unity Profiler performance data including CPU usage, memory statistics, GPU frame time, and call hierarchy analysis";
            Uri = "unity://profiler/{dataType}";
            IsAsync = false;
            
            InitializeProfilerRecorders();
        }

        /// <summary>
        /// Initialize ProfilerRecorders for various performance metrics
        /// </summary>
        private void InitializeProfilerRecorders()
        {
            try
            {
                // CPU metrics
                _profilerRecorders["MainThreadFrameTime"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "Main Thread", 15);
                _profilerRecorders["CPUTotalFrameTime"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "CPU Total Frame Time", 15);
                _profilerRecorders["CPURenderThreadFrameTime"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "CPU Render Thread Frame Time", 15);
                
                // Memory metrics  
                _profilerRecorders["SystemUsedMemory"] = ProfilerRecorder.StartNew(ProfilerCategory.Memory, "System Used Memory");
                _profilerRecorders["GCReservedMemory"] = ProfilerRecorder.StartNew(ProfilerCategory.Memory, "GC Reserved Memory");
                _profilerRecorders["GCUsedMemory"] = ProfilerRecorder.StartNew(ProfilerCategory.Memory, "GC Used Memory");
                
                // GPU metrics
                _profilerRecorders["GPUFrameTime"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "GPU Frame Time", 15);
                
                // Update/Physics metrics
                _profilerRecorders["BehaviourUpdate"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "BehaviourUpdate", 15);
                _profilerRecorders["LateBehaviourUpdate"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "LateBehaviourUpdate", 15);
                _profilerRecorders["FixedBehaviourUpdate"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "FixedBehaviourUpdate", 15);
                _profilerRecorders["PhysicsFixedUpdate"] = ProfilerRecorder.StartNew(ProfilerCategory.Internal, "FixedUpdate.PhysicsFixedUpdate", 15);
                
                // Rendering metrics (Editor only)
#if UNITY_EDITOR
                _profilerRecorders["DrawCallsCount"] = ProfilerRecorder.StartNew(ProfilerCategory.Render, "Draw Calls Count");
                _profilerRecorders["BatchesCount"] = ProfilerRecorder.StartNew(ProfilerCategory.Render, "Batches Count");
                _profilerRecorders["SetPassCallsCount"] = ProfilerRecorder.StartNew(ProfilerCategory.Render, "SetPass Calls Count");
                _profilerRecorders["TrianglesCount"] = ProfilerRecorder.StartNew(ProfilerCategory.Render, "Triangles Count");
                _profilerRecorders["VerticesCount"] = ProfilerRecorder.StartNew(ProfilerCategory.Render, "Vertices Count");
#endif
            }
            catch (Exception ex)
            {
                Debug.LogError($"[MCP Unity] Failed to initialize ProfilerRecorders: {ex.Message}");
            }
        }

        /// <summary>
        /// Fetch profiler data based on the requested data type
        /// </summary>
        /// <param name="parameters">Resource parameters containing dataType filter</param>
        /// <returns>A JObject containing profiler performance data</returns>
        public override JObject Fetch(JObject parameters)
        {
            try
            {
                string dataType = parameters?["dataType"]?.ToObject<string>()?.ToLower() ?? "all";
                
                JObject result = new JObject
                {
                    ["success"] = true,
                    ["timestamp"] = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss.fff"),
                    ["frameCount"] = Time.frameCount
                };

                switch (dataType)
                {
                    case "cpu":
                        result["data"] = GetCPUData();
                        break;
                    case "memory":
                        result["data"] = GetMemoryData();
                        break;
                    case "gpu":
                        result["data"] = GetGPUData();
                        break;
                    case "rendering":
                        result["data"] = GetRenderingData();
                        break;
                    case "hierarchy":
                        result["data"] = GetHierarchyData();
                        break;
                    case "summary":
                        result["data"] = GetPerformanceSummary();
                        break;
                    case "all":
                    default:
                        result["data"] = GetAllProfilerData();
                        break;
                }

                return result;
            }
            catch (Exception ex)
            {
                Debug.LogError($"[MCP Unity] Error fetching profiler data: {ex.Message}");
                return new JObject
                {
                    ["success"] = false,
                    ["message"] = $"Failed to fetch profiler data: {ex.Message}",
                    ["error"] = "profiler_error"
                };
            }
        }

        /// <summary>
        /// Get CPU performance data
        /// </summary>
        private JObject GetCPUData()
        {
            return new JObject
            {
                ["mainThreadFrameTime"] = GetRecorderAverage("MainThreadFrameTime"),
                ["cpuTotalFrameTime"] = GetRecorderAverage("CPUTotalFrameTime"),
                ["cpuRenderThreadFrameTime"] = GetRecorderAverage("CPURenderThreadFrameTime"),
                ["behaviourUpdate"] = GetRecorderAverage("BehaviourUpdate"),
                ["lateBehaviourUpdate"] = GetRecorderAverage("LateBehaviourUpdate"),
                ["fixedBehaviourUpdate"] = GetRecorderAverage("FixedBehaviourUpdate"),
                ["physicsFixedUpdate"] = GetRecorderAverage("PhysicsFixedUpdate"),
                ["fps"] = Time.frameCount > 0 ? 1.0f / Time.unscaledDeltaTime : 0f
            };
        }

        /// <summary>
        /// Get memory performance data
        /// </summary>
        private JObject GetMemoryData()
        {
            return new JObject
            {
                ["systemUsedMemory"] = GetRecorderValue("SystemUsedMemory") / (1024 * 1024), // MB
                ["gcReservedMemory"] = GetRecorderValue("GCReservedMemory") / (1024 * 1024), // MB
                ["gcUsedMemory"] = GetRecorderValue("GCUsedMemory") / (1024 * 1024), // MB
                ["totalAllocatedMemory"] = UnityEngine.Profiling.Profiler.GetTotalAllocatedMemoryLong() / (1024 * 1024), // MB
                ["totalReservedMemory"] = UnityEngine.Profiling.Profiler.GetTotalReservedMemoryLong() / (1024 * 1024), // MB
                ["totalUnusedReservedMemory"] = UnityEngine.Profiling.Profiler.GetTotalUnusedReservedMemoryLong() / (1024 * 1024) // MB
            };
        }

        /// <summary>
        /// Get GPU performance data
        /// </summary>
        private JObject GetGPUData()
        {
            return new JObject
            {
                ["gpuFrameTime"] = GetRecorderAverage("GPUFrameTime")
            };
        }

        /// <summary>
        /// Get rendering performance data
        /// </summary>
        private JObject GetRenderingData()
        {
            JObject renderData = new JObject();
            
#if UNITY_EDITOR
            renderData["drawCalls"] = GetRecorderValue("DrawCallsCount");
            renderData["batches"] = GetRecorderValue("BatchesCount");
            renderData["setPassCalls"] = GetRecorderValue("SetPassCallsCount");
            renderData["triangles"] = GetRecorderValue("TrianglesCount");
            renderData["vertices"] = GetRecorderValue("VerticesCount");
#else
            renderData["note"] = "Rendering metrics are only available in Unity Editor";
#endif
            
            return renderData;
        }

        /// <summary>
        /// Get CPU call hierarchy data for the current frame
        /// </summary>
        private JObject GetHierarchyData()
        {
            JObject hierarchyData = new JObject();
            
#if UNITY_EDITOR
            try
            {
                int currentFrame = ProfilerDriver.lastFrameIndex;
                if (currentFrame >= 0)
                {
                    using (var frameData = ProfilerDriver.GetHierarchyFrameDataView(currentFrame, 0, 
                        HierarchyFrameDataView.ViewModes.Default, HierarchyFrameDataView.columnTotalTime, false))
                    {
                        if (frameData.valid)
                        {
                            hierarchyData["frameIndex"] = currentFrame;
                            hierarchyData["threadName"] = frameData.threadName;
                            hierarchyData["frameTimeMs"] = frameData.frameTimeMs;
                            
                            JArray topCalls = new JArray();
                            int rootId = frameData.GetRootItemID();
                            
                            _parentsCacheList.Clear();
                            frameData.GetItemDescendantsThatHaveChildren(rootId, _parentsCacheList);
                            
                            foreach (int parentId in _parentsCacheList)
                            {
                                _childrenCacheList.Clear();
                                frameData.GetItemChildren(parentId, _childrenCacheList);
                                
                                if (_childrenCacheList.Count > 0)
                                {
                                    JObject callInfo = new JObject
                                    {
                                        ["name"] = frameData.GetItemName(parentId),
                                        ["totalTimeMs"] = frameData.GetItemColumnDataAsFloat(parentId, HierarchyFrameDataView.columnTotalTime),
                                        ["selfTimeMs"] = frameData.GetItemColumnDataAsFloat(parentId, HierarchyFrameDataView.columnSelfTime),
                                        ["calls"] = frameData.GetItemColumnDataAsFloat(parentId, HierarchyFrameDataView.columnCalls),
                                        ["gcMemory"] = frameData.GetItemColumnDataAsFloat(parentId, HierarchyFrameDataView.columnGcMemory),
                                        ["childrenCount"] = _childrenCacheList.Count
                                    };
                                    
                                    topCalls.Add(callInfo);
                                }
                            }
                            
                            // Sort by total time descending and take top 10
                            var sortedCalls = new List<JObject>();
                            foreach (JObject call in topCalls)
                            {
                                sortedCalls.Add(call);
                            }
                            sortedCalls.Sort((a, b) => 
                                b["totalTimeMs"].ToObject<float>().CompareTo(a["totalTimeMs"].ToObject<float>()));
                            
                            JArray top10Calls = new JArray();
                            for (int i = 0; i < Math.Min(10, sortedCalls.Count); i++)
                            {
                                top10Calls.Add(sortedCalls[i]);
                            }
                            
                            hierarchyData["topCostlyMethods"] = top10Calls;
                        }
                        else
                        {
                            hierarchyData["error"] = "Frame data not valid";
                        }
                    }
                }
                else
                {
                    hierarchyData["error"] = "No valid frame data available";
                }
            }
            catch (Exception ex)
            {
                hierarchyData["error"] = $"Failed to get hierarchy data: {ex.Message}";
            }
#else
            hierarchyData["note"] = "CPU call hierarchy is only available in Unity Editor";
#endif
            
            return hierarchyData;
        }

        /// <summary>
        /// Get a performance summary with key metrics and recommendations
        /// </summary>
        private JObject GetPerformanceSummary()
        {
            JObject summary = new JObject();
            
            float mainThreadTime = GetRecorderAverage("MainThreadFrameTime");
            float gpuTime = GetRecorderAverage("GPUFrameTime");
            float currentFPS = Time.frameCount > 0 ? 1.0f / Time.unscaledDeltaTime : 0f;
            long gcMemoryMB = GetRecorderValue("GCReservedMemory") / (1024 * 1024);
            
            summary["currentFPS"] = currentFPS;
            summary["mainThreadTimeMs"] = mainThreadTime;
            summary["gpuTimeMs"] = gpuTime;
            summary["gcMemoryMB"] = gcMemoryMB;
            
            // Performance assessment
            JArray issues = new JArray();
            JArray recommendations = new JArray();
            
            if (currentFPS < 30)
            {
                issues.Add("Low FPS detected");
                recommendations.Add("Consider optimizing CPU/GPU performance");
            }
            
            if (mainThreadTime > 16.67f) // 60 FPS target
            {
                issues.Add("High CPU frame time");
                recommendations.Add("Profile CPU usage and optimize expensive operations");
            }
            
            if (gpuTime > 16.67f)
            {
                issues.Add("High GPU frame time");
                recommendations.Add("Optimize rendering, reduce draw calls or shader complexity");
            }
            
            if (gcMemoryMB > 100)
            {
                issues.Add("High GC memory usage");
                recommendations.Add("Reduce managed allocations and implement object pooling");
            }
            
            summary["issues"] = issues;
            summary["recommendations"] = recommendations;
            summary["performanceGrade"] = GetPerformanceGrade(currentFPS, mainThreadTime, gpuTime);
            
            return summary;
        }

        /// <summary>
        /// Get all profiler data in a comprehensive format
        /// </summary>
        private JObject GetAllProfilerData()
        {
            return new JObject
            {
                ["cpu"] = GetCPUData(),
                ["memory"] = GetMemoryData(),
                ["gpu"] = GetGPUData(),
                ["rendering"] = GetRenderingData(),
                ["summary"] = GetPerformanceSummary()
            };
        }

        /// <summary>
        /// Get average value from a ProfilerRecorder
        /// </summary>
        private float GetRecorderAverage(string recorderName)
        {
            if (!_profilerRecorders.TryGetValue(recorderName, out var recorder) || !recorder.Valid)
                return 0f;

            var samplesCount = recorder.Capacity;
            if (samplesCount == 0)
                return 0f;

            double sum = 0;
            var samples = new List<ProfilerRecorderSample>(samplesCount);
            recorder.CopyTo(samples);
            for (var i = 0; i < samples.Count; ++i)
                sum += samples[i].Value;

            return (float)(sum / samples.Count / 1000000.0); // Convert to milliseconds
        }

        /// <summary>
        /// Get current value from a ProfilerRecorder
        /// </summary>
        private long GetRecorderValue(string recorderName)
        {
            if (!_profilerRecorders.TryGetValue(recorderName, out var recorder) || !recorder.Valid)
                return 0L;

            return recorder.LastValue;
        }

        /// <summary>
        /// Calculate performance grade based on key metrics
        /// </summary>
        private string GetPerformanceGrade(float fps, float cpuTime, float gpuTime)
        {
            if (fps >= 60 && cpuTime <= 16.67f && gpuTime <= 16.67f)
                return "A"; // Excellent
            else if (fps >= 45 && cpuTime <= 22.22f && gpuTime <= 22.22f)
                return "B"; // Good
            else if (fps >= 30 && cpuTime <= 33.33f && gpuTime <= 33.33f)
                return "C"; // Fair
            else if (fps >= 20)
                return "D"; // Poor
            else
                return "F"; // Very Poor
        }

        /// <summary>
        /// Cleanup ProfilerRecorders when object is disposed
        /// </summary>
        ~GetProfilerResource()
        {
            foreach (var recorder in _profilerRecorders.Values)
            {
                if (recorder.Valid)
                    recorder.Dispose();
            }
        }
    }
} 